### PR TITLE
add input/output for custom model

### PR DESCRIPTION
**Pull request by ylwu-amzn**
Signed-off-by: Yaliang Wu <ylwu@amazon.com>

### Description
add input/output for custom model
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/src/main/java/org/opensearch/ml/common/input/MLInput.java [hunk 1]
```diff
@@ -40,6 +42,12 @@ public class MLInput implements Input {
     public static final String INPUT_INDEX_FIELD = "input_index";
     public static final String INPUT_QUERY_FIELD = "input_query";
     public static final String INPUT_DATA_FIELD = "input_data";
+    // For trained model
+    public static final String RETURN_BYTES_FIELD = "return_bytes";
+    public static final String RETURN_NUMBER_FIELD = "return_number";
+    public static final String TARGET_RESPONSE_FIELD = "target_response";
+    public static final String TARGET_RESPONSE_POSITIONS_FIELD = "target_response_positions";
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/input/MLInput.java:49]: These three fields requires description imo.

**Review by Zhangxunmt — APPROVED** [review]

File: common/src/main/java/org/opensearch/ml/common/MLTaskType.java [hunk 2]
```diff
@@ -9,5 +9,7 @@ public enum MLTaskType {
     TRAINING,
     PREDICTION,
     TRAINING_AND_PREDICTION,
-    EXECUTION
+    EXECUTION,
+    UPLOAD_MODEL,
+    LOAD_MODEL
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/MLTaskType.java:14]: What about UNLOAD_MODEL? are you planning to add later?

File: common/src/main/java/org/opensearch/ml/common/exception/MLException.java [hunk 3]
```diff
@@ -32,6 +32,15 @@ public MLException(Throwable cause) {
         super(cause);
     }
 
+    /**
+     * Constructor with specified error message adn cause.
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/exception/MLException.java:36]: typo -> and

File: common/src/main/java/org/opensearch/ml/common/input/MLInput.java [hunk 4]
```diff
@@ -140,6 +168,12 @@ public static MLInput parse(XContentParser parser, String inputAlgoName) throws
         List<String> sourceIndices = new ArrayList<>();
         DataFrame dataFrame = null;
 
+        boolean returnBytes = false;
+        boolean returnNumber = true;
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/input/MLInput.java:172]: why this is initialized as `true`?

File: common/src/main/java/org/opensearch/ml/common/output/model/MLResultDataType.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+public enum MLResultDataType {
+    FLOAT32(Format.FLOATING, 4),
+    FLOAT64(Format.FLOATING, 8),
+    FLOAT16(Format.FLOATING, 2),
+    UINT8(Format.UINT, 1),
+    INT32(Format.INT, 4),
+    INT8(Format.INT, 1),
+    INT64(Format.INT, 8),
+    BOOLEAN(Format.BOOLEAN, 1),
+    UNKNOWN(Format.UNKNOWN, 0),
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/output/model/MLResultDataType.java:17]: Why do we need this `unknown` format. Will we be able to parse this type?

File: common/src/main/java/org/opensearch/ml/common/output/model/ModelTensors.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                output.writeTo(out);
+            }
+        } else {
+            out.writeBoolean(false);
+        }
+    }
+
+    public void filter(ModelResultFilter resultFilter) {
+        boolean returnBytes = resultFilter.isReturnBytes();
+        boolean returnNUmber = resultFilter.isReturnNumber();
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/output/model/ModelTensors.java:69]: returnNUmber -> returnNumber?

**Review by dhrubo-os — APPROVED** [review]

File: common/src/main/java/org/opensearch/ml/common/MLTaskType.java (hunk 2 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/MLTaskType.java:14]: Actually unload model part is quite lightweight (just remove some objects from memory), I think we don't need to create ML task for it. Creating a new unload task may add too much overhead comparing the simple unload action.

File: common/src/main/java/org/opensearch/ml/common/input/MLInput.java (hunk 4 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/input/MLInput.java:172]: By default we should only return number for embeddings. If set it as `false` , no embedding vectors will return.

File: common/src/main/java/org/opensearch/ml/common/output/model/MLResultDataType.java (hunk 5 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/output/model/MLResultDataType.java:17]: There will be some gap between model output data type and java type. For such data types not recognized, we will set it as UNKNOWN.

File: common/src/main/java/org/opensearch/ml/common/input/MLInput.java (hunk 1 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/input/MLInput.java:49]: Sure, will add some description in next PR.

File: common/src/main/java/org/opensearch/ml/common/output/model/ModelTensors.java (hunk 6 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/output/model/ModelTensors.java:69]: As this PR has been approved. I will address these typos in next PR.

**Pull request merged**